File: solutions/Next-Permutation/Next-Permutation.py


class Solution:
    # @param num, a list of integer
    # @return a list of integer
    def nextPermutation(self, num):
        n=len(num)
        mark=-1
        for i in range(n-1,0,-1):
            if num[i]<=num[i-1]:
                continue
            else:
                mark=i-1
                break
        if mark==-1:
            num.reverse()
            return num
        for i in range(n-1,mark,-1):
            if num[i]>num[mark]:
                mark2=i
                break
        num[mark],num[mark2]=num[mark2],num[mark]
        num2=num[mark+1:]
        num2.reverse()
        for i in range(mark+1,n):
            num[i]=num2[i-mark-1]
        return num
 

File: solutions/Next-Permutation/Next-Permutation.cpp

// class Solution {
// public:
//     void nextPermutation(vector<int> &num) {
//         int n=num.size();
//         int mark=-1;
//         for(int i=n-1;i>=0;i--)
//         {
//             if(num[i]<=num[i-1])
//                 continue;
//             else
//             {
//                 mark=i-1;
//                 break;
//             }
//         }
//         if(mark==-1)
//         {
//             reverse(num.begin(),num.end());
//             return;
//         }
//         int mark2=0;
//         for(int i=n-1;i>mark;i--)
//         {
//             if(num[i]>num[mark])
//             {
//                 mark2=i;
//                 break;
//             }
//         }
//         swap(num[mark],num[mark2]);
//         reverse(num.begin()+mark+1,num.end());
//         return ;
//     }
// };

// 2019-07-20
// O(n)
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int index1 = -1;
        for(int i=n-1;i>0;i--)
        {
            if(nums[i-1] < nums[i])
            {
                index1 = i-1;
                break;
            }
        }
        
        if (index1 == -1)
        {
            reverse(nums.begin(), nums.end());
        }
        else
        {
            int index2 = n-1;
            for(int i=n-1;i>index1;i--)
            {
                if(nums[i] > nums[index1])
                {
                    index2 = i;
                    break;
                }
            }
            swap(nums[index1], nums[index2]);
            reverse(nums.begin() + index1 + 1, nums.end());
        }
    }
private:
    void swap(int &x, int &y)
    {
        int tmp = x;
        x = y;
        y = tmp;
    }
};
 